File: GoInterpreter/parser/parser.go
package parser

import (
	. "GoInterpreter/node"
	"fmt"
	"os"
)

///////////////////////////////////////////////////////////////////////////////
// Mapping Nonterminal Nodes
///////////////////////////////////////////////////////////////////////////////

var NonTerminalsMap map[string]*Node = map[string]*Node{
	// Completed lines:
	"CODE-BLOCK CODE-BLOCK": BuildNewNonterm("CODE-BLOCK", 0, 0, "g", "BOTH", [2]int{0, 1}, []*Node{}),
	"ASGN-STMT":             BuildNewNonterm("CODE-BLOCK", 0, 0, "g", "LEFT", [2]int{0, 0}, []*Node{}),
	// VAL (logical value) assignments and operations:
	"DEF VAL-VAR AS VAL-EXPR ;":       BuildNewNonterm("ASGN-STMT", 0, 0, "g", "assignVal", [2]int{1, 3}, []*Node{}),
	"VAL ID":                          BuildNewNonterm("VAL-VAR", 0, 0, "g", "idAssert", [2]int{1, 1}, []*Node{}),
	"NOT VAL-SUBEXPR":                 BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildNot", [2]int{1, 1}, []*Node{}),
	"NOT VAL-VAR":                     BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildNot", [2]int{1, 1}, []*Node{}),
	"( VAL-SUBEXPR OR VAL-SUBEXPR )":  BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildOr", [2]int{1, 3}, []*Node{}),
	"( VAL-VAR OR VAL-SUBEXPR )":      BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildOr", [2]int{1, 3}, []*Node{}),
	"( VAL-SUBEXPR OR VAL-VAR )":      BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildOr", [2]int{1, 3}, []*Node{}),
	"( VAL-VAR OR VAL-VAR )":          BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildOr", [2]int{1, 3}, []*Node{}),
	"( VAL-SUBEXPR AND VAL-SUBEXPR )": BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildAnd", [2]int{1, 3}, []*Node{}),
	"( VAL-VAR AND VAL-SUBEXPR )":     BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildAnd", [2]int{1, 3}, []*Node{}),
	"( VAL-SUBEXPR AND VAL-VAR )":     BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildAnd", [2]int{1, 3}, []*Node{}),
	"( VAL-VAR AND VAL-VAR )":         BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "buildAnd", [2]int{1, 3}, []*Node{}),
	"( VAL-VAR )":                     BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "LEFT", [2]int{1, 1}, []*Node{}),
	"( VAL-SUBEXPR )":                 BuildNewNonterm("VAL-SUBEXPR", 0, 0, "g", "LEFT", [2]int{1, 1}, []*Node{}),
	"VAL-SUBEXPR OR VAL-SUBEXPR":      BuildNewNonterm("VAL-EXPR", 0, 0, "g", "buildOr", [2]int{0, 2}, []*Node{}),
	"VAL-VAR OR VAL-SUBEXPR":          BuildNewNonterm("VAL-EXPR", 0, 0, "g", "buildOr", [2]int{0, 2}, []*Node{}),
	"VAL-SUBEXPR OR VAL-VAR":          BuildNewNonterm("VAL-EXPR", 0, 0, "g", "buildOr", [2]int{0, 2}, []*Node{}),
	"VAL-VAR OR VAL-VAR":              BuildNewNonterm("VAL-EXPR", 0, 0, "g", "buildOr", [2]int{0, 2}, []*Node{}),
	"VAL-SUBEXPR AND VAL-SUBEXPR":     BuildNewNonterm("VAL-EXPR", 0, 0, "g", "buildAnd", [2]int{0, 2}, []*Node{}),
	"VAL-VAR AND VAL-SUBEXPR":         BuildNewNonterm("VAL-EXPR", 0, 0, "g", "buildAnd", [2]int{0, 2}, []*Node{}),
	"VAL-SUBEXPR AND VAL-VAR":         BuildNewNonterm("VAL-EXPR", 0, 0, "g", "buildAnd", [2]int{0, 2}, []*Node{}),
	"VAL-VAR AND VAL-VAR":             BuildNewNonterm("VAL-EXPR", 0, 0, "g", "buildAnd", [2]int{0, 2}, []*Node{}),
	"VAL-SUBEXPR":                     BuildNewNonterm("VAL-EXPR", 0, 0, "g", "LEFT", [2]int{0, 0}, []*Node{}),
}

///////////////////////////////////////////////////////////////////////////////
// Recursive, Iterative Parser
///////////////////////////////////////////////////////////////////////////////

func extractNodeNames(fromNodes []*Node) string {
	var nodeNames string = ""
	for n := range fromNodes {
		nodeNames += fromNodes[n].Name + " "
	}
	//fmt.Printf("'%s'\n", nodeNames[:len(nodeNames)-1])
	return nodeNames[:len(nodeNames)-1]
}

func idAssert(intoTokens []*Node, withID *Node) []*Node {
	fmt.Printf("idAssert TRIGGERED FOR %s!\n", withID.Left.Literal)
	var updatedTokens []*Node
	for _, node := range intoTokens {
		if node.Name == "ID" && node.Literal == withID.Left.Literal {
			var name string = withID.Name
			var start int = node.Start
			var end int = node.End
			var scope string = node.Scope
			var xidNode *Node = BuildNewNonterm(name, start, end, scope, "LEFT", [2]int{0, 0}, []*Node{node})
			updatedTokens = append(updatedTokens, xidNode)
		} else {
			updatedTokens = append(updatedTokens, node)
		}
	}
	fmt.Println("*" + extractNodeNames(intoTokens))
	return updatedTokens
}

func Parse(lejTokens []*Node) *Node {
	for j := 5; j > 0; j-- {
		var i int = 0
		for i < (len(lejTokens)-j)+1 {
			var tokSlice []*Node = lejTokens[i : i+j]
			var tokNames string = extractNodeNames(tokSlice)
			var newNode *Node
			if nonterm, ok := NonTerminalsMap[tokNames]; ok {
				if nonterm.Children[0] == nonterm.Children[1] { // This production rule only one Left Node.
					newNode = BuildNewNonterm(nonterm.Name, 0, 0, "?", nonterm.Action, nonterm.Children, tokSlice)
					fmt.Printf("%v\n", newNode)
					var s int = newNode.Left.Start
					var e int = newNode.Left.End
					var scope string = newNode.Left.Scope
					newNode = BuildNewNonterm(newNode.Name, s, e, scope, newNode.Action, newNode.Children, tokSlice)
					fmt.Printf("%v\n", newNode)
					fmt.Printf("L: %s -> %s :: %s\n", newNode.Name, newNode.Left.Name, tokNames)
				} else { // This production rule has a Left Node and a Right Node.
					newNode = BuildNewNonterm(nonterm.Name, 0, 0, "?", nonterm.Action, nonterm.Children, tokSlice)
					fmt.Printf("%v\n", newNode)
					var s int = newNode.Left.Start
					var e int = newNode.Right.End
					var scope string = newNode.Left.Scope
					newNode = BuildNewNonterm(newNode.Name, s, e, scope, newNode.Action, newNode.Children, tokSlice)
					fmt.Printf("%v\n", newNode)
					fmt.Printf("LR: %s -> %s %s :: %s\n", newNode.Name, newNode.Left.Name, newNode.Right.Name, tokNames)
				}
				lejTokens[i] = newNode
				// idAssert actions occur mid-parse.
				if newNode.Action == "idAssert" {
					lejTokens = idAssert(lejTokens, newNode)
				}
				copy(lejTokens[i+1:], lejTokens[i+j:])
				lejTokens = lejTokens[:(len(lejTokens)-j)+1]
				fmt.Printf("**%s\n\n", extractNodeNames(lejTokens))
				return Parse(lejTokens)
			} else {
				i += 1
			}
		}
	}
	if len(lejTokens) > 1 {
		// Panic, because the parser couldn't complete the parse.
		fmt.Printf("This collection of tokens cannot be parsed.\n")
		os.Exit(1)
	}
	return lejTokens[0]
}

func PrintParse(fromToks []*Node) *Node {
	var lejTree *Node = Parse(fromToks)
	fmt.Printf("%v\n", lejTree)
	return lejTree
}

File: PythonInterpreter/parser_.py
'''
This module parses Lej code into an AST bottom-up,
according to the generation rules provided herein.
'''
from node import Node, build_new_nonterm

###############################################################################
# Mapping Nonterminal Nodes
###############################################################################

NONTERMINALS_MAP: dict[str, Node] = \
    {  # Completed lines:
    'CODE-BLOCK CODE-BLOCK': \
    build_new_nonterm('CODE-BLOCK', 0, 0, 'g', 'BOTH', (0, 1), []),
    'ASGN-STMT': \
    build_new_nonterm('CODE-BLOCK', 0, 0, 'g', 'LEFT', (0, 0), []),
    # VAL (logical value) assignments and operations:
    'DEF VAL-VAR AS VAL-EXPR ;': \
    build_new_nonterm('ASGN-STMT', 0, 0, 'g', 'assign_val', (1, 3), []),
    'VAL ID': \
    build_new_nonterm('VAL-VAR', 0, 0, 'g', 'id_assert', (1, 1), []),
    'NOT VAL-SUBEXPR': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_not', (1, 1), []),
    'NOT VAL-VAR': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_not', (1, 1), []),
    '( VAL-SUBEXPR OR VAL-SUBEXPR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_or', (1, 3), []),
    '( VAL-VAR OR VAL-SUBEXPR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_or', (1, 3), []),
    '( VAL-SUBEXPR OR VAL-VAR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_or', (1, 3), []),
    '( VAL-VAR OR VAL-VAR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_or', (1, 3), []),
    '( VAL-SUBEXPR AND VAL-SUBEXPR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_and', (1, 3), []),
    '( VAL-VAR AND VAL-SUBEXPR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_and', (1, 3), []),
    '( VAL-SUBEXPR AND VAL-VAR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_and', (1, 3), []),
    '( VAL-VAR AND VAL-VAR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'build_and', (1, 3), []),
    '( VAL-VAR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'LEFT', (1, 1), []),
    '( VAL-SUBEXPR )': \
    build_new_nonterm('VAL-SUBEXPR', 0, 0, 'g', 'LEFT', (1, 1), []),
    'VAL-SUBEXPR OR VAL-SUBEXPR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'build_or', (0, 2), []),
    'VAL-VAR OR VAL-SUBEXPR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'build_or', (0, 2), []),
    'VAL-SUBEXPR OR VAL-VAR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'build_or', (0, 2), []),
    'VAL-VAR OR VAL-VAR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'build_or', (0, 2), []),
    'VAL-SUBEXPR AND VAL-SUBEXPR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'build_and', (0, 2), []),
    'VAL-VAR AND VAL-SUBEXPR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'build_and', (0, 2), []),
    'VAL-SUBEXPR AND VAL-VAR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'build_and', (0, 2), []),
    'VAL-VAR AND VAL-VAR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'build_and', (0, 2), []),
    'VAL-SUBEXPR': \
    build_new_nonterm('VAL-EXPR', 0, 0, 'g', 'LEFT', (0, 0), [])}

###############################################################################
# Recursive, Iterative Parser
###############################################################################


def extract_node_names(from_nodes: list[Node]) -> str:
    '''
    summary: Extract the name attribute from `from_nodes` and return them
        as a space-separated string.
    '''
    return ' '.join(node.name for node in from_nodes)


def id_assert(into_tokens: list[Node], with_id: Node) -> list[Node]:
    '''
    summary: Check if there is an ID in `into_tokens`
        with the same literal value as the left child of `with_id`,
        and build a new nonterminal `VAL-VAR` node using that ID node if found.
    '''
    assert isinstance(with_id.left, Node)
    updated_tokens: list[Node] = []
    for node in into_tokens:
        if node.name == 'ID' and node.literal == with_id.left.literal:
            name = with_id.name
            start = node.start
            end = node.end
            scope = node.scope
            xid_node = build_new_nonterm(name, start, end, scope,
                                         'LEFT', (0, 0), [node])
            updated_tokens.append(xid_node)
        else:
            updated_tokens.append(node)
    return updated_tokens


def parse(lej_tokens: list[Node]) -> Node:
    '''
    summary: Iteratively check `lej_tokens` for `NONERMINALS_MAP` matches
        and replace the matched sublist with the corresponding Node
        and the relevant children from the matched sublist.
        Continue this until `lej_tokens` only one nonterminal Node remains,
        and return that Node.

        Along the way, reassert any found `<TYPE>-ID` Node
            as the parent Node type of terminal Nodes with the same ID literal.
    '''
    for j in range(5, 0, -1):
        i = 0
        while i < (len(lej_tokens) - j) + 1:
            tok_slice = lej_tokens[i:i+j]
            tok_names = extract_node_names(tok_slice)
            new_node = None
            if tok_names in NONTERMINALS_MAP:
                nonterm: Node = NONTERMINALS_MAP[tok_names]
                # This production rule only one Left Node.
                if nonterm.children[0] == nonterm.children[1]:
                    new_node = build_new_nonterm(nonterm.name, 0, 0, "?",
                                                 nonterm.action,
                                                 nonterm.children, tok_slice)
                    assert isinstance(new_node.left, Node)
                    s = new_node.left.start
                    e = new_node.left.end
                    scope = new_node.left.scope
                    new_node = build_new_nonterm(new_node.name, s, e, scope,
                                                 new_node.action,
                                                 new_node.children, tok_slice)
                    print(f'L: {new_node.name} -> {new_node.left.name} '
                          f':: {tok_names}',)
                else:  # This production rule has a Left Node and a Right Node.
                    new_node = build_new_nonterm(nonterm.name, 0, 0, "?",
                                                 nonterm.action,
                                                 nonterm.children, tok_slice)
                    assert isinstance(new_node.left, Node)
                    assert isinstance(new_node.right, Node)
                    s = new_node.left.start
                    e = new_node.right.end
                    scope = new_node.left.scope
                    new_node = build_new_nonterm(new_node.name, s, e, scope,
                                                 new_node.action,
                                                 new_node.children, tok_slice)
                    print(f'LR: {new_node.name} -> {new_node.left.name} '
                          f'{new_node.right.name} :: {tok_names}',)

                lej_tokens[i] = new_node
                # id_assert actions occur mid-parse.
                if new_node.action == "id_assert":
                    lej_tokens = id_assert(lej_tokens, new_node)
                print("BEFORE:", i, j, extract_node_names(lej_tokens))
                lej_tokens[i+1:] = lej_tokens[i+j:]
                print("AFTER:", i, j, extract_node_names(lej_tokens))
                return parse(lej_tokens)
            else:
                i += 1
    if len(lej_tokens) > 1:
        # Panic, because the parser couldn't complete the parse.
        print("This collection of tokens cannot be parsed.")
        exit(1)
    return lej_tokens[0]
